apis/Google.Cloud.Spanner.Data/Google.Cloud.Spanner.Data/Keys.cs

```diff
@@ -0,0 +1,319 @@
+    {
+        /// <summary>
+        /// True for key sets that select all keys in the table or index, and false for all other key sets.
+        /// </summary>
+        public bool AllKeys { get; private set; }
+
+        /// <summary>
+        /// The collection of specific keys to select.
+        /// </summary>
+        public IEnumerable<Key> Keys { get; }
+
+        /// <summary>
+        /// The list of key ranges to select. All rows that are in at least one of the key ranges will be selected.
+        /// </summary>
+        public IEnumerable<KeyRange> Ranges { get; }
+
+        private Lazy<V1.KeySet> LazyProtobuf => new Lazy<V1.KeySet>(ToProtobuf);
+        internal V1.KeySet Protobuf => LazyProtobuf.Value;
+
+        /// <summary>
+        /// Returns a key set that selects all keys in the table or index.
+        /// </summary>
+        public static KeySet All => new KeySet { AllKeys = true };
+
+        /// <summary>
+        /// Creates a new key set from a collection of parameters that represent exactly one key.
+        /// </summary>
+        /// <param name="parameters">A parameter collection containing the values of the key</param>
+        /// <returns>A key set for one specific key</returns>
+        public static KeySet FromParameters(SpannerParameterCollection parameters) => FromKeys(new Key(parameters));
+
+        /// <summary>
+        /// Creates a new key set for a collection of keys.
+        /// </summary>
+        /// <param name="keys">The keys to select</param>
+        /// <returns>A key set for a collection of keys</returns>
+        public static KeySet FromKeys(params Key[] keys) => new KeySet(GaxPreconditions.CheckNotNull(keys, nameof(keys)));
+
+        /// <summary>
+        /// Creates a new key set for a collection of key ranges.
+        /// </summary>
+        /// <param name="ranges">The key ranges to select</param>
+        /// <returns>A key set for a collection of key ranges</returns>
+        public static KeySet FromRanges(params KeyRange[] ranges) => new KeySet(GaxPreconditions.CheckNotNull(ranges, nameof(ranges)));
+
+        private KeySet(IEnumerable<Key> keys)
+        {
+            Keys = new List<Key>(GaxPreconditions.CheckNotNull(keys, nameof(keys))).AsReadOnly();
+        }
+
+        private KeySet(IEnumerable<KeyRange> ranges)
+        {
+            Ranges = new List<KeyRange>(GaxPreconditions.CheckNotNull(ranges, nameof(ranges))).AsReadOnly();
+        }
+
+        private KeySet()
+        {
+        }
+
+        /// <inheritdoc/>
+        public override bool Equals(object other)
+        {
+            if (other is KeySet keySet)
+            {
+                return Protobuf.Equals(keySet.Protobuf);
+            }
+            return false;
+        }
+
+        /// <inheritdoc/>
+        public override int GetHashCode()
+        {
+            return Protobuf.GetHashCode();
+        }
+
+        private V1.KeySet ToProtobuf()
+        {
+            if (AllKeys)
```

Comment:
Although Spanner does support key sets that are a combination of for example a collection of keys and a collection of key ranges, this wrapper class only accepts one type. So a `KeySet` is always either:
- AllKeys
- A collection of `Keys`
- A collection of `KeyRanges`

And never a combination of any of the above. This is done to keep it simple, while still supporting all use cases (a `KeyRange` can always be specified in such a way that it is a single key).

Reply:
But can't you just build a `V1.KeySet` with everything set on this `KeySet`? I mean what is the complexity that you are trying to avoid? Or am I missing something.

Reply:
Yes, creating it from everything in this `KeySet` would not be the problem. The simplification lays in the API that we offer to the users:
- A `KeySet` instance can only be created through one of the factory methods, and these only accept either a collection of `Key`s, a collection of `KeyRange`s or an `AllKeys` instance. This makes sure that only a 'simple' KeySet instance is created.
- This prevents any questions/confusion regarding combinations like 'what happens if I have a set of Keys and a set of KeyRanges? What happens if I have a KeySet that sets both AllKeys=true and a number of Keys? While those questions are relatively simple to answer, there is also no real need to ever create a KeySet like that, which the current implementation also does not allow the creation of it.

Reply:
Ah yes, I see. Sounds good. I suppose that if we get the demand for it (after all it can be done through the API), we can always add more methods that combine different key sets. But agreed that we can start with this for now.